File: src/84-largest-rectangle-in-histogram.py
# 84. Largest Rectangle in Histogram

class Solution:
    def largestRectangleArea(self, heights: 'List[int]') -> 'int':
        if len(heights) == 0:
            return 0
        first_lower_left = [0] * len(heights)  # index of the first bar that is lower than heights[i] on its left
        first_lower_right = [0] * len(heights) # index of the first bar that is lower than heights[i] on its right
        first_lower_left[0] = -1
        first_lower_right[len(heights) - 1] = len(heights)

        for i in range(1, len(heights)):
            idx = i - 1
            while idx >= 0 and heights[idx] >= heights[i]:
                idx = first_lower_left[idx]
            first_lower_left[i] = idx

        for i in range(len(heights) - 2, -1, -1):
            idx = i + 1
            while idx < len(heights) and heights[idx] >= heights[i]:
                idx = first_lower_right[idx]
            first_lower_right[i] = idx

        max_area = 0
        for i in range(0, len(heights)):
            max_area = max(max_area, (first_lower_right[i] - first_lower_left[i] - 1) * heights[i])
        return max_area

        ## TLE
        # for i in range(0, len(heights)):
        #     li = i - 1
        #     while li >= 0 and heights[li] >= heights[i]:
        #         li -= 1
        #     edge_from_left = li
        #     ri = i + 1
        #     while ri < len(heights) and heights[ri] >= heights[i]:
        #         ri += 1
        #     edge_from_right = ri
        #     max_area = max(max_area, (edge_from_right - edge_from_left - 1) * heights[i])
        # return max_area


if __name__ == "__main__":
    import time
    start = time.clock()

    s = Solution()
    ans = s.largestRectangleArea([2,1,5,6,2,3])
    print(ans)

    end = time.clock()
    print('Running time: %s ms' % ((end - start) * 1000))
        

File: src/84-largest-rectangle-in-histogram.cpp
// 84. Largest Rectangle in Histogram


#include <iostream>
#include <vector>

using namespace std;

// O(N) in 3 pass.
// https://leetcode.com/problems/largest-rectangle-in-histogram/discuss/28902/5ms-O(n)-Java-solution-explained-(beats-96)
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
		int n = heights.size();
		if (n == 0) {
			return 0;
		}
        vector<int> leftBound(n);  // "leftBound[i] = j" means the right most j with heights[j] < heights[i] and j < i
		leftBound[0] = -1;
        vector<int> rightBound(n);  // "rightBound[i] = j" means the left most j with heights[j] < heights[i] and j > i
		rightBound[n - 1] = n;

		// this block is actually O(n)
		// every index will only be pointed to (or, referenced) once.
		for (int i = 1; i < n; ++i) {
			int bound = i - 1;
			while (bound >= 0 && heights[bound] >= heights[i]) {
				bound = leftBound[bound];
			}
			leftBound[i] = bound;
		}

		for (int i = n - 2; i >= 0; --i) {
			int bound = i + 1;
			while (bound < n && heights[bound] >= heights[i]) {
				bound = rightBound[bound];
			}
			rightBound[i] = bound;
		}

		int maxArea = 0;
		for (int i = 0; i < n; ++i) {
			maxArea = max(maxArea, (rightBound[i] - leftBound[i] - 1) * heights[i]);
		}
		return maxArea;
    }
};

// // Divide and Conquer O(nlogn). NOT accepted in worst case
// class Solution {
// public:
//     int largestRectangleArea(vector<int>& heights) {
//         return rectangleBetween(heights, 0, heights.size() - 1);
//     }
// private:
//     int rectangleBetween(const vector<int>& heights, int left, int right) {
//         if (left > right) {
//             return 0;
//         }
//         int shortestIndex = left;
//         for (int i = left + 1; i <= right; ++i) {
//             if (heights[i] < heights[shortestIndex]) {
//                 shortestIndex = i;
//             }
//         }
//         int areaWithShortest = heights[shortestIndex] * (right - left + 1);
//         int areaLeft = rectangleBetween(heights, left, shortestIndex - 1);
//         int areaRight = rectangleBetween(heights, shortestIndex + 1, right);
//         return max(areaWithShortest, max(areaLeft, areaRight));
//     }
// };
